framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,563 @@
+        for (Map.Entry<String, Object> f : a.fieldValues.entrySet()) {
+            AnnotationFieldType aft = a.def().fieldTypes.get(f.getKey());
+            sj.add(f.getKey() + "=" + formatAnnotationValue(aft, f.getValue()));
+        }
+        return sj.toString();
+    }
+
+    /**
+     * Returns all annotations in {@code annos} in a form suitable to be printed as Java source
+     * code.
+     *
+     * <p>Each annotation is followed by a space, to separate it from following Java code.
+     *
+     * @param annos the annotations to format
+     * @return all annotations in {@code annos}, separated by spaces, in a form suitable to be
+     *     printed as Java source code
+     */
+    private static String formatAnnotations(Collection<? extends Annotation> annos) {
+        StringBuilder sb = new StringBuilder();
+        for (Annotation tla : annos) {
+            if (!isInternalJDKAnnotation(tla.def.name)) {
+                sb.append(formatAnnotation(tla));
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Formats the component types of an array via recursive descent through the array's scene-lib
+     * structure.
+     *
+     * @param e the array's scenelib type element
+     * @param arrayType the string representation of the array's type
+     * @return the type formatted to be written to Java source code
+     */
+    private static String formatArrayType(ATypeElement e, String arrayType) {
+        StringBuilder result = new StringBuilder();
+        return formatArrayTypeImpl(e, arrayType, result);
+    }
+
+    /**
+     * The implementation of formatArrayType. Because of the (insane) structure of scenelib, the
+     * descent order is innermost array type to outermost array type, then component type. So if we
+     * have the type {@code @Foo int @Bar [] @Baz []}, the iteration order is {@code @Bar []}, then
+     * {@code @Baz []}, and then finally {@code @Foo int}. This implementation therefore passes a
+     * string builder with the result of the array types seen so far, to handle multidimensional
+     * arrays. That builder is appended to the final component type in the base case.
+     *
+     * @param e same as above
+     * @param arrayType same as above
+     * @param result the string builder containing the array types seen so far
+     * @return the formatted string, as above
+     */
+    private static String formatArrayTypeImpl(
+            ATypeElement e, String arrayType, StringBuilder result) {
+        String nextComponentType =
+                arrayType.indexOf('[') == -1
+                        ? null
+                        : arrayType.substring(0, arrayType.lastIndexOf('['));
+        // base case when the component is a non-array type
+        if (nextComponentType == null) {
+            return formatAnnotations(e.tlAnnotationsHere) + arrayType + " " + result.toString();
```

Comment:
Does result contain a trailing space if it's nonempty?

Reply:
Yes. I updated the documentation to note that